src/popthelp.c

```diff
@@ -67,7 +67,7 @@ struct poptOption poptHelpOptions[] = {
 } ;
 
 static struct poptOption poptHelpOptions2[] = {
-  { NULL, '\0', POPT_ARG_INTL_DOMAIN, PACKAGE, 0, NULL, NULL},
+  { NULL, '\0', POPT_ARG_INTL_DOMAIN, (void *)PACKAGE, 0, NULL, NULL},
```

Comment:
This doesn't really fix it, just silences a warning which is often worse. Why not add eg
```
static const char *intl_domain = PACKAGE;
```
...and point poptHelpOptions2 to that?

Reply:
`struct poptOption` takes a `void *` as `arg`, because it can takes several different types of values dependent on the value of `argInfo`.
Also void-casting this struct member is common, as the next line contains: `(void *)displayArgs`

> static const char *intl_domain = PACKAGE;
This won't help since the variable is still specified const.

Reply:
> This won't help since the variable is still specified const.
Actually it does, because "const char *" gives you a pointer whose value you can freely change, you just can't modify the data it points to. You don't even need that cast with it.

Reply:
```
popthelp.c:72:39: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   72 |   { NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
      |                                       ^~~~~~~~~~~
popthelp.c:72:39: error: initializer element is not constant
popthelp.c:72:39: note: (near initialization for ‘poptHelpOptions2[0].arg’)
```
It's not about the pointer value, its about the pointed to memory.

A way would be to create a new editable string: `char intl_domain[] = PACKAGE;`

But we already depend on the implementation doing only the things it should with this `void *` based on the preceding argument (here `POPT_ARG_INTL_DOMAIN`).
Everywhere where `if (poptArgType(opt) == POPT_ARG_INTL_DOMAIN)` is used, `opt->arg` is converted into `const char *`.

Reply:
What compiler is this? gcc 10.2.1 with -Wwrite-strings stops complaining when with my suggestion (I did actually try it...)

Reply:
gcc (Debian 10.2.0-6) 10.2.0
Debian clang version 11.0.0-+rc2-5

Code:
```
{ NULL, '\0', POPT_ARG_INTL_DOMAIN, PACKAGE, 0, NULL, NULL},
```
Output (gcc):
```
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..  -I. -I.. -DPOPT_SYSCONFDIR="\"/usr/local/etc\""   -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c -o popthelp.lo popthelp.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I. -I.. -DPOPT_SYSCONFDIR=\"/usr/local/etc\" -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c popthelp.c  -fPIC -DPIC -o .libs/popthelp.o
In file included from system.h:6,
                 from popthelp.c:11:
../config.h:114:17: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  114 | #define PACKAGE "popt"
      |                 ^~~~~~
popthelp.c:70:39: note: in expansion of macro ‘PACKAGE’
   70 |   { NULL, '\0', POPT_ARG_INTL_DOMAIN, PACKAGE, 0, NULL, NULL},
      |                                       ^~~~~~~
```
Output (clang):
```
/bin/bash ../libtool  --tag=CC   --mode=compile clang-11 -DHAVE_CONFIG_H -I. -I..  -I. -I.. -DPOPT_SYSCONFDIR="\"/usr/local/etc\""   -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c -o popthelp.lo popthelp.c
libtool: compile:  clang-11 -DHAVE_CONFIG_H -I. -I.. -I. -I.. -DPOPT_SYSCONFDIR=\"/usr/local/etc\" -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c popthelp.c  -fPIC -DPIC -o .libs/popthelp.o
popthelp.c:70:39: warning: initializing 'void *' with an expression of type 'const char [5]' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  { NULL, '\0', POPT_ARG_INTL_DOMAIN, PACKAGE, 0, NULL, NULL},
                                      ^~~~~~~
../config.h:114:17: note: expanded from macro 'PACKAGE'
#define PACKAGE "popt"
                ^~~~~~
1 warning generated.
```

Code:
```
static const char *intl_domain = PACKAGE;
...
{ NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
```
Output (gcc):
```
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..  -I. -I.. -DPOPT_SYSCONFDIR="\"/usr/local/etc\""   -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c -o popthelp.lo popthelp.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I. -I.. -DPOPT_SYSCONFDIR=\"/usr/local/etc\" -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c popthelp.c  -fPIC -DPIC -o .libs/popthelp.o
popthelp.c:71:39: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   71 |   { NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
      |                                       ^~~~~~~~~~~
popthelp.c:71:39: error: initializer element is not constant
popthelp.c:71:39: note: (near initialization for ‘poptHelpOptions2[0].arg’)
```
Output (clang):
```
/bin/bash ../libtool  --tag=CC   --mode=compile clang-11 -DHAVE_CONFIG_H -I. -I..  -I. -I.. -DPOPT_SYSCONFDIR="\"/usr/local/etc\""   -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c -o popthelp.lo popthelp.c
libtool: compile:  clang-11 -DHAVE_CONFIG_H -I. -I.. -I. -I.. -DPOPT_SYSCONFDIR=\"/usr/local/etc\" -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c popthelp.c  -fPIC -DPIC -o .libs/popthelp.o
popthelp.c:71:39: warning: initializing 'void *' with an expression of type 'const char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  { NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
                                      ^~~~~~~~~~~
popthelp.c:71:39: error: initializer element is not a compile-time constant
  { NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
                                      ^~~~~~~~~~~
1 warning and 1 error generated.
```

Code:
```
static const char *const intl_domain = PACKAGE;
...
{ NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
```
Output (gcc):
```
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..  -I. -I.. -DPOPT_SYSCONFDIR="\"/usr/local/etc\""   -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c -o popthelp.lo popthelp.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -I. -I.. -DPOPT_SYSCONFDIR=\"/usr/local/etc\" -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c popthelp.c  -fPIC -DPIC -o .libs/popthelp.o
popthelp.c:71:39: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   71 |   { NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
      |                                       ^~~~~~~~~~~
```
Output (clang):
```
/bin/bash ../libtool  --tag=CC   --mode=compile clang-11 -DHAVE_CONFIG_H -I. -I..  -I. -I.. -DPOPT_SYSCONFDIR="\"/usr/local/etc\""   -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c -o popthelp.lo popthelp.c
libtool: compile:  clang-11 -DHAVE_CONFIG_H -I. -I.. -I. -I.. -DPOPT_SYSCONFDIR=\"/usr/local/etc\" -Wwrite-strings -MT popthelp.lo -MD -MP -MF .deps/popthelp.Tpo -c popthelp.c  -fPIC -DPIC -o .libs/popthelp.o
popthelp.c:71:39: warning: initializing 'void *' with an expression of type 'const char *const' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  { NULL, '\0', POPT_ARG_INTL_DOMAIN, intl_domain, 0, NULL, NULL},
                                      ^~~~~~~~~~~
1 warning generated.
```

Reply:
Well this is most peculiar. Previously when I tried (several times), the intl_domain approach silenced the warning, and didn't see anything like the above. And now that I tried once again, I'm getting the same as you - it's not merely a warning it's an error. I guess I did run autoreconf in between but fail to see how that'd change such a thing, but that's the only *change* I can think of.

Oh well, sorry about the noise but when you see totally different behavior on your system, you have to ask questions...